hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/ReopenTableRegionsProcedure.java

```diff
@@ -139,33 +170,57 @@ protected Flow executeFromState(MasterProcedureEnv env, ReopenTableRegionsState
       case REOPEN_TABLE_REGIONS_CONFIRM_REOPENED:
         regions = regions.stream().map(env.getAssignmentManager().getRegionStates()::checkReopened)
           .filter(l -> l != null).collect(Collectors.toList());
-        if (regions.isEmpty()) {
-          return Flow.NO_MORE_STATE;
+        // we need to create a set of region names because the HRegionLocation hashcode is only
+        // based
+        // on the server name
+        Set<byte[]> currentRegionBatchNames = currentRegionBatch.stream()
+          .map(r -> r.getRegion().getRegionName()).collect(Collectors.toSet());
+        currentRegionBatch = regions.stream()
+          .filter(r -> currentRegionBatchNames.contains(r.getRegion().getRegionName()))
+          .collect(Collectors.toList());
+        if (currentRegionBatch.isEmpty()) {
+          if (regions.isEmpty()) {
+            return Flow.NO_MORE_STATE;
+          } else {
+            setNextState(ReopenTableRegionsState.REOPEN_TABLE_REGIONS_REOPEN_REGIONS);
+            if (reopenBatchBackoffMillis > 0) {
+              backoff(reopenBatchBackoffMillis);
+            }
+            return Flow.HAS_MORE_STATE;
+          }
         }
-        if (regions.stream().anyMatch(loc -> canSchedule(env, loc))) {
+        if (currentRegionBatch.stream().anyMatch(loc -> canSchedule(env, loc))) {
```

Comment:
I think the changes in this method are sort of confusing. I think the division of labor should be:

`REOPEN_TABLE_REGIONS_REOPEN_REGIONS` -- breaks the set of `regions` into a smaller batch, and reopens that batch. The code you have for that looks ok
`REOPEN_TABLE_REGIONS_CONFIRM_REOPENED` -- checks if any regions are still needing reopen. If so, go back to REOPEN_REGIONS. This is how it used to work.

If we totally reverted the changes in REOPEN_TABLE_REGIONS_CONFIRM_REOPENED, I think we'd have successfully broken the full region list into batches and the flow would work, just without backoff. So in this method we really just need to add the backoff.

The backoff as you have it below looks ok at first glance. This is just a long winded way of saying we should drop the highlighted code :)

Reply:
I think this complexity is necessary in order to differentiate between `are there more regions to reopen` vs `are there more regions to reopen in our current batch`, one being a state which is considered a retry and exponentially increasing backoff, the other being a state change and fixed backoff

Reply:
On second thought, after a discussion on https://issues.apache.org/jira/browse/HBASE-25549, i wonder if we should take the safer approach where require each batch to finish before scheduling more. This feature could serve a dual purpose of progressive rollout and rate limiting.

We could also trivially update this code to do an actual progressive deploy -- first batch size 1, then 2, then 4, etc up to the current batch size config you added. At that point it stays at that max concurrency until completion.

Thoughts on adding that? I think we just need one more field for `currentBatchSize` which we increment after each batch up to `reopenBatchSize`. It might make sense to persist this new field in the proto.

Reply:
> Now let's look at your new code. I see what you are saying. You are basically doing:
If currentRegionBatch are all reopened, exit or process the next batch
Otherwise, can any of the remaining be scheduled? If not, exponential backoff. If so, process next batch.

I think that's missing a few distinct outcomes of `CONFIRM_REOPENED`, and that's where this complexity comes into play. Here are the possibilities:
1. The current batch has reopened successfully and there are no more regions to produce a next batch, so finish
2. The current batch has reopened successfully and there are more regions to produce a next batch, so set the next state to `REOPEN_REGIONS` and suspend. This will process the next batch
3. The current batch has not reopened successfully, and some of the current batch can still be scheduled. In this case return to `REOPEN_REGIONS` to avoid an infinite loop due to a bypassed proc or something. This could technically reprocess the next batch, or redefine what the "current batch" looks like, but I think that's ok given the exceptional nature of this case
4. The current batch has not reopened successfully, and none of the current batch can still be scheduled (because they have already been scheduled). In this case we're just waiting on TRSPs to complete, so retain current state of `CONFIRM_REOPENED` and retry w/ increasing backoff 

If I just revert this diff as described then this procedure will just get stuck in infinite loops (that I think could theoretically be exited if TRSP finishes quickly enough?) and all unit tests fail

> We could also trivially update this code to do an actual progressive deploy -- first batch size 1, then 2, then 4, etc up to the current batch size config you added. At that point it stays at that max concurrency until completion.

I like this idea, and I think it would be trivial to add in the current state of this PR. We'd just need to store a maxReopenBatchSize and increase the reopenBatchSize to Math.min(2*reopenBatchSize, maxReopenBatchSize) each time we enter state 2 above

Reply:
Ok lets table the missing outcomes/simplification discussion in that case, since we'll want to keep the current logic for the progressive reopen. We'll also want to do as i said, where in CONFIRM we update currentRegionBatch to filter out the reopened regions, and then in REOPEN we should only create a new currentRegionBatch if the existing isEmpty.

Once we have that in there, I need to take another look at the logic here. I think it could be a bit clearer, but don't want to make any suggestions until we have the complete impl in place.

In terms of progressive, it'd be nice to start with 1 region but otherwise agree with your Math.min idea.

Reply:
Sounds good, I just pushed a change which does the following:
1. each reopen proc will start with a batch size of 1
2. the batch size will double after each confirmed batch (until we hit our max)
3. our tests confirm the intended batching behavior, and that each region is only opened once